subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -44,58 +44,108 @@ type HivedAlgorithm struct {
 	vcSchedulers map[api.VirtualClusterName]intraVCScheduler
 	// scheduler for opportunistic pods
 	opportunisticSchedulers map[CellChain]*topologyAwareScheduler
-	// ChainCellLists of physical cells of each cell chain
+	// ChainCellLists of physical cells of each cell chain (including the children of the free cells)
 	fullCellList map[CellChain]ChainCellList
 	// ChainCellLists of free physical cells of each cell chain (used in buddy alloc)
 	freeCellList map[CellChain]ChainCellList
+	// all affinity groups that have been allocated cells
+	allocatedAffinityGroups map[string]*AlgoAffinityGroup
+
+	// vcFreeCellNum, allVCFreeCellNum, and totalLeftCellNum are used to track cell usage of the VCs.
+	// Note that these numbers count both healthy and bad cells.
+
+	// number of free preassigned cells of each VC of each cell type
+	vcFreeCellNum map[api.VirtualClusterName]map[CellChain]map[CellLevel]int32
+	// total number of preassigned free cells in all the VCs of each cell type
+	allVCFreeCellNum map[CellChain]map[CellLevel]int32
+	// number of cells left in the physical cluster of each cell type
+	// (i.e., the cell is either in the free list or could be obtained by splitting such a free cell.)
+	totalLeftCellNum map[CellChain]map[CellLevel]int32
+
+	// number of bad free cells in the physical cluster of each cell type
+	badFreeCellNum map[CellChain]map[CellLevel]int32
+	// bad nodes in the physical cluster
+	badNodes common.Set
 	// map each GPU type to all chains that contain this type
 	chains map[string][]CellChain
 	// map each level in a chain to the specific cell type name
 	cellTypes map[CellChain]map[CellLevel]api.CellType
-	// all affinity groups that have been allocated cells
-	allocatedAffinityGroups map[string]*AlgoAffinityGroup
-	// all reserved physical cells (VC -> reservation ID -> cells)
-	reservedCells map[api.VirtualClusterName]map[api.ReservationId]*PhysicalCell
+	// cluster status exposed to external
+	apiClusterStatus api.ClusterStatus
 	// lock
 	algorithmLock sync.RWMutex
 }
 
-// NewHivedAlgorithm initializes a HivedAlgorithm from the config file
+// NewHivedAlgorithm initializes a HivedAlgorithm from the config file.
 func NewHivedAlgorithm(sConfig *api.Config) *HivedAlgorithm {
-	pcl, gpuNums, gpuTypeToChain, cellLevelToType, nonReservedVcl, reservedVcl, reservedPc := ParseConfig(sConfig)
+	fullPcl, freePcl, vcQuotas, nonReservedFullVcl, nonReservedFreeVcl, reservedVcl, reservedPc,
```

Comment:
>vcQuotas [](start = 19, length = 8)

Rename quota to be a more concert concept. Such as total/free cells. Also for all places.